src/devices/vdml_usd.cpp

```diff
@@ -21,5 +21,90 @@ std::int32_t is_installed(void) {
 	return usd_is_installed();
 }
 
+std::int32_t list_files_raw(const char* path, char* buffer, int32_t len) {
+	return usd_list_files_raw(path, buffer, len);
+}
+
+std::vector<std::string> list_files(const char* path) {
+	std::vector<std::string> files = {};
+	// malloc buffer to store file names
+	size_t buffer_size = 10000;
+	char *buffer = (char *) malloc(buffer_size);
+	if (buffer == NULL) {
+		// try again smaller buffer to see if that works
+		buffer_size = 500;
+		buffer = (char *) malloc(buffer_size);
+		if (buffer == NULL) {
+			// if still fails, return vector containing error state
+			// set errno to ENOMEM
+			errno = ENOMEM;
+			files.push_back("ERROR");
+			files.push_back("not enough memory to get file names");
+			return files;
+		}
+	}
+
+	// Check path user passed in
+	std::string_view path_sv(path);
+	size_t found = path_sv.find("usd");
+	if (found == 0 || found == 1) {
+		// Deal with when user prepends path with usd
+		// as either "usd/..." or "/usd/..."
+		path_sv.remove_prefix(3);
+	}
+
+	// set path to path_sv.data()
+	path = path_sv.data();
+
+	// Call the C function
+	int32_t success = usd_list_files_raw(path, buffer, buffer_size);
+	// Check if call successful, if error return vector containing error state
+	if (success == PROS_ERR) {
+		files.push_back("ERROR");
+		// Check errno to see which error state occurred
+		// push back error state to files vector as std::string
+		if (errno == EINVAL || errno == ENOENT) {
+			// errors related to path not found
+			files.push_back("path not found");
+		} else {
+			// other error stats related to file io
+			files.push_back("file i/o error");
+		}
+		return files;
+	}
+
+	// Parse buffer given call successful, split by '/n'
+	std::string_view str(buffer);
+
+	// delimter_pos is the position of the delimiter '\n'
+	// index of which character to start substr from
+	// file_name used to store each file name
+	size_t delimiter_pos = 0;
+	size_t index = 0;
+	std::string_view file_name;
+
+	// Loop until delimiter '\n' can not be found anymore
+	while ((delimiter_pos = str.find('\n', index)) != std::string::npos) {
```

Comment:
If we really wanted to optimize this even more you could write a sentinel function (or whatever people call it I'm tired atm) that grabs all the delimter indexes, and then another loop that stores all the delimter positions in a data structure such as a vector, and then another loop that iterates through the data structure. This changes the O(N * M) behavior of this to O(N + M) where N is the size of the string and M is the number of delimeters.

Reply:
I'm not sure where you got `O(N*M)`.  If I'm reading correctly, you're seeing it as a full loop through all N characters each time we loop? That's not true, though, as `index` changes where the iteration starts each time (starting further along the string on each iteration), and each iteration stops as soon as it locates `'\n'`. Therefore, this is only O(N).

Reply:
> I'm not sure where you got `O(N*M)`. If I'm reading correctly, you're seeing it as a full loop through all N characters each time we loop? That's not true, though, as `index` changes where the iteration starts each time (starting further along the string on each iteration), and each iteration stops as soon as it locates `'\n'`. Therefore, this is only O(N).

ah you're right, I didn't notice the index parameter last night.